vm/src/org/mozartoz/truffle/nodes/call/CallMethodNode.java

```diff
@@ -43,89 +46,130 @@ public static CallMethodNode create() {
-			@Cached("methodFromArgs(self, args)") OzProc cachedProc,
-			@Cached("createDirectCallNode(cachedProc.callTarget)") DirectCallNode callNode) {
-		assert args.length == 1;
-		Object[] arguments = new Object[] { self, args[0] };
-		return callNode.call(frame, OzArguments.pack(cachedProc.declarationFrame, arguments));
+		Lookup lookup = lookupNode.executeLookup(self, name);
+		return dispatchNode.executeDispatch(frame, self, lookup, args);
 	}
 
-	@Specialization(guards = {
-			"self.getClazz() == cachedClazz",
-			"labelFromArgs(args) == cachedLabel",
-			"methodNotFound",
-	}, limit = "methodCache(3)")
-	protected Object callOtherwise(VirtualFrame frame, OzObject self, Object[] args,
-			@Cached("self.getClazz()") DynamicObject cachedClazz,
-			@Cached("labelFromArgs(args)") Object cachedLabel,
-			@Cached("methodFromArgs(self, args) == null") boolean methodNotFound,
-			@Cached("getMethod(self, OTHERWISE)") OzProc otherwiseProc,
-			@Cached("createDirectCallNode(otherwiseProc.callTarget)") DirectCallNode callNode) {
-		assert args.length == 1;
-		Object[] arguments = new Object[] { self, OTHERWISE_MESSAGE_FACTORY.newRecord(args[0]) };
-		return callNode.call(frame, OzArguments.pack(otherwiseProc.declarationFrame, arguments));
-	}
+	@ImportStatic(CallMethodNode.class)
+	@NodeChildren({ @NodeChild("self"), @NodeChild("name") })
+	public abstract static class MethodLookupNode extends OzNode {
 
-	@Specialization(contains = { "callClassLabelIdentity", "callProcIdentity", "callOtherwise" })
-	protected Object callObject(VirtualFrame frame, OzObject self, Object[] args,
-			@Cached("create()") IndirectCallNode callNode,
-			@Cached("createBinaryProfile()") ConditionProfile otherwise) {
-		assert args.length == 1;
-		Object message = derefNode.executeDeref(args[0]);
+		public static MethodLookupNode create() {
+			return MethodLookupNodeGen.create(null, null);
+		}
 
-		final Object name = labelNode.executeLabel(message);
-		assert OzGuards.isLiteral(name);
+		public abstract Lookup executeLookup(OzObject self, Object name);
+
+		@Specialization(guards = {
+				"self.getClazz() == cachedClazz",
+				"name == cachedName",
+		}, limit = "methodCache(3)")
+		protected Lookup descriptorIdentityLookup(OzObject self, Object name,
+				@Cached("self.getClazz()") DynamicObject cachedClazz,
+				@Cached("name") Object cachedName,
+				@Cached("genericLookup(self, name)") Lookup cachedLookup) {
+			return cachedLookup;
+		}
+
+		BranchProfile otherwiseProfile = BranchProfile.create();
+
+		@Specialization(contains = "descriptorIdentityLookup")
+		protected Lookup genericLookup(OzObject self, Object name) {
+			OzProc proc = getMethod(self, name);
+			if (proc == null) {
+				otherwiseProfile.enter();
+				proc = getMethod(self, OTHERWISE);
+				if (proc == null) {
+					CompilerDirectives.transferToInterpreterAndInvalidate();
+					throw new RuntimeException(name + " not found");
```

Comment:
Such a handling is probably fine, although it would be best to tell the class name.
OTOH, maybe it's impossible to encounter this case since the base class already has an otherwise method?
Then we would only need an `assert != null`.